compiler/rustc_trait_selection/src/traits/project.rs

```diff
@@ -1070,6 +1070,40 @@ fn assemble_candidates_from_impls<'cx, 'tcx>(
                         | ty::Infer(..)
                         | ty::Error(_) => false,
                     }
+                } else if lang_items.async_destruct_trait() == Some(trait_ref.def_id) {
+                    match self_ty.kind() {
+                        ty::Bool
+                        | ty::Char
+                        | ty::Int(_)
+                        | ty::Uint(_)
+                        | ty::Float(_)
+                        | ty::Adt(..)
+                        | ty::Str
+                        | ty::Array(..)
+                        | ty::Slice(_)
+                        | ty::RawPtr(..)
+                        | ty::Ref(..)
+                        | ty::FnDef(..)
+                        | ty::FnPtr(..)
+                        | ty::Dynamic(..)
+                        | ty::Closure(..)
+                        | ty::CoroutineClosure(..)
+                        | ty::Coroutine(..)
+                        | ty::CoroutineWitness(..)
+                        | ty::Never
+                        | ty::Tuple(..)
+                        | ty::Infer(ty::InferTy::IntVar(_) | ty::InferTy::FloatVar(..)) => true,
+
+                         // type parameters, opaques, and unnormalized projections have pointer
+                        // metadata if they're known (e.g. by the param_env) to be sized
```

Comment:
copy paste from above, and the preexisting above comment is also wrong, as it's for `discriminant_kind`, not `pointee`

Reply:
I am not sure what comment should I put there rn.

Reply:
for `discriminant_kind`, you can say that these types don't have a known discriminant and may need to be normalized further or rely on param env for discriminant projections.

here, the same thing, but for `AsyncDestruct`

Reply:
It's not in the scope of this PR, so I've made another one https://github.com/rust-lang/rust/pull/122377